src/libraries/System.Private.DataContractSerialization/src/System/Xml/XmlStreamNodeWriter.cs

```diff
@@ -344,37 +335,54 @@ protected unsafe void UnsafeWriteUnicodeChars(char* chars, int charCount)
 
         protected unsafe int UnsafeGetUnicodeChars(char* chars, int charCount, byte[] buffer, int offset)
         {
-            char* charsMax = chars + charCount;
-            while (chars < charsMax)
+            if (BitConverter.IsLittleEndian)
             {
-                char value = *chars++;
-                buffer[offset++] = (byte)value;
-                value >>= 8;
-                buffer[offset++] = (byte)value;
+                new ReadOnlySpan<byte>((byte*)chars, 2 * charCount)
+                    .CopyTo(buffer.AsSpan(offset));
             }
+            else
+            {
+                char* charsMax = chars + charCount;
+                while (chars < charsMax)
+                {
+                    char value = *chars++;
+                    buffer[offset++] = (byte)value;
+                    buffer[offset++] = (byte)(value >> 8);
+                }
+            }
+
             return charCount * 2;
         }
 
         protected unsafe int UnsafeGetUTF8Length(char* chars, int charCount)
         {
             char* charsMax = chars + charCount;
-            while (chars < charsMax)
+
+            // This method is only called from 2 places and will use length of at least (128/3 and 256/3) respectivly
+            // We avoid Vector<T> sine it is unsure how downclocking due to AVX512 would affect total throughput
+            if (Vector256.IsHardwareAccelerated
+                && Vector256<short>.Count < charCount && charCount <= 2048)
             {
-                if (*chars >= 0x80)
-                    break;
+                char* lastSimd = chars + charCount - Vector256<short>.Count;
+                Vector256<short> mask = Vector256.Create(unchecked((short)0xff80));
 
-                chars++;
-            }
+                while (chars < lastSimd)
+                {
+                    if (((*(Vector256<short>*)chars) & mask) != Vector256<short>.Zero)
+                        goto NonAscii;
 
-            if (chars == charsMax)
-                return charCount;
+                    chars += Vector256<short>.Count;
+                }
 
-            char[] chArray = new char[charsMax - chars];
-            for (int i = 0; i < chArray.Length; i++)
-            {
-                chArray[i] = chars[i];
+                if ((*(Vector256<short>*)lastSimd & mask) == Vector256<short>.Zero)
+                    return charCount;
             }
-            return (int)(chars - (charsMax - charCount)) + GetByteCount(chArray);
+
+        NonAscii:
+            int numRemaining = (int)(charsMax - chars);
+            int numAscii = charCount - numRemaining;
+
+            return numAscii + (_encoding ?? s_UTF8Encoding).GetByteCount(chars, numRemaining);
```

Comment:
What are the possible values of `_encoding`? Can it be something other than Utf8?

Note that it better to call `Encoding.UTF8.GetBytes` directly without caching the encoding locally. `Encoding.UTF8.GetBytes` allows devitalization optimization to kick in that eliminates the overhead of Encoding being an abstract type.

Reply:
It can be passed by the user when creating a text XmlDictionaryWriter,  but it is only set to _encoding if the codepage is the same as utf8. 
So in theory it can be any encoding class even if unlikely .

for `s_encoding` it does not use the default constructor but passes is `(false, true)` so I did no dare to do that change.
If it does not change the behaviour then that can be a simple follow up fix.